### PR TITLE
allow ContainerResource calculations to continue with missing metrics like Resource calculations

### LINKED ISSUE #127169: HPA with container metrics fails when any pod is not in a ready state
### What happened?

We've noticed that our HPA will stay at it's current scale and not make decisions when a pod is in an un-Ready state. IE a single pod is in CrashLoopBackoff and the HPA shows:

```
...
resource cpu of container "mycontainer" on pods  (as a percentage of request):     <unknown> / 50%
...
Warning  FailedGetResourceMetric  78s (x7 over 4m7s)  horizontal-pod-autoscaler  unable to get metric cpu: failed to get container metrics: container guides3 not present in metrics for pod namesapce/mycontainer-20240904t1338-433ca6a1-db998df74-s5rfh
```

That particular pod was in CrashLoopBackoff at the time and it wasn't until i removed it was it able to make scaling decisions

We're also running istio so there is another container in these pods and it's running ok; this may be why it's showing up in the metrics but unable to get info on the main container

### What did you expect to happen?

I would expect the HPA to ignore pods in this state and use only the Ready pods for its decisions

### How can we reproduce it (as minimally and precisely as possible)?

Setup an HPA with container based metrics
Put one of the pods in bad state, could just have a single pod consume a lot of memory and have it OOM a few times
See the HPA make no decisions while it's in this state

### Anything else we need to know?

I've looked through the code and it seems to be steming from:
`metrics, err := c.client.PodMetricses(namespace).List(ctx, metav1.ListOptions{LabelSelector: selector.String()})`
This client just uses the selector and doesn't care about the state any of the pods might be in. It does end up filtering out pods later in the reconcile cycle but this returns an error here if it's not in the list of metrics because a single container may not be running 

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.29.6
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.6-gke.1326000
```

</details>

### Cloud provider

<details>
GCP
</details>

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by DP19**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

Container metrics based HPA's fail to compute any decision when there is missing metrics for a container within a pod. This can happen when a pod is not in a ready status (has OOM'd or crashed and is in a CrashLoopBackOff, or being terminated and one of the containers has exited and another hasn't etc).
During these windows we should still make decisions in the HPA cycle since we filter out these pods after getting these metrics, currently we just error and exit the loop when this happens

This enables parity between the Pod based metrics and Container based which is what I think most people expect to happen.

#### Which issue(s) this PR fixes:
Fixes #127169

#### Does this PR introduce a user-facing change?
```release-note
HPA's with ContainerResource metrics will no longer error when container metrics are missing, instead they will use the same logic Resource metrics are using to make calculations
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

### CODE REVIEW

File: pkg/controller/podautoscaler/metrics/client.go
Comment (raywainman) [inline pkg/controller/podautoscaler/metrics/client.go:78]: Could we document this new behavior in the interface here?

https://github.com/kubernetes/kubernetes/blob/0c5e832aa2bb32573ca695858bc421d111c90119/pkg/controller/podautoscaler/metrics/interfaces.go#L41

File: pkg/controller/podautoscaler/metrics/client.go
Comment (raywainman) [inline pkg/controller/podautoscaler/metrics/client.go:104]: nit: what do you think about making this a warning?

This isn't an expected steady-state thing so perhaps we can make this more visible in the logs incase someone isn't realizing a container is crash looping?

File: pkg/controller/podautoscaler/metrics/client_test.go
Comment (raywainman) [inline pkg/controller/podautoscaler/metrics/client_test.go:432]: Can we add a test for the case where we are returning metrics for a container that is missing?

File: pkg/controller/podautoscaler/replica_calculator_test.go
Comment (raywainman) [inline pkg/controller/podautoscaler/replica_calculator_test.go:449]: What do you think about converting this test to one that returns a proper replica number when the container is missing from some pods?

File: pkg/controller/podautoscaler/metrics/client.go
Comment (DP19) [reply pkg/controller/podautoscaler/metrics/client.go:104]: I don't really have a strong opinion here other than I used the same log level we are currently using in getPodMetrics.

https://github.com/kubernetes/kubernetes/blob/0c5e832aa2bb32573ca695858bc421d111c90119/pkg/controller/podautoscaler/metrics/client.go#L122

If we do update it, we should probably do it in both places

Comment (DP19) [reply pkg/controller/podautoscaler/metrics/client.go:104]: It also looks like this package only has `info` and `error`. I can adjust the verbosity int here; but will probably need a pointer to what each levels mean in this repo

File: pkg/controller/podautoscaler/metrics/client_test.go
Comment (DP19) [reply pkg/controller/podautoscaler/metrics/client_test.go:432]: So this is that test, we just don't expect it to error now. I could change the case here to look for the container that is actually missing in the metrics and check that it isn't returned if that's what we're looking for

Comment (DP19) [reply pkg/controller/podautoscaler/metrics/client_test.go:432]: If anything, i think this test actually highlights the underlying bug here. The container in the test case we're asking for metrics on is `test-1` and both of the pods here have metrics for that container; its just one of them that is missing metrics for the second container and producing the error; for the container we didn't even ask for metrics on

File: pkg/controller/podautoscaler/replica_calculator_test.go
Comment (DP19) [reply pkg/controller/podautoscaler/replica_calculator_test.go:449]: Works for me; there's a similar test here and I just added in another one that includes metrics for one container in two pods but not the one we're asking for

**Review by raywainman — COMMENTED** [review]
/ok-to-test

/lgtm

**Comment by raywainman** [comment]
@sanposhiho - would you be able to take a look at this?

**Comment by sanposhiho** [comment]
-1 
At least, it requires more discussion.

/hold

> Warning  FailedGetResourceMetric  78s (x7 over 4m7s)  horizontal-pod-autoscaler  unable to get metric cpu: failed to get container metrics: container guides3 not present in metrics for pod namesapce/mycontainer-20240904t1338-433ca6a1-db998df74-s5rfh

This part from the issue. Did you really use the container level metrics? AFAIK, the container level metric shouldn't emit this event. (`FailedGetContainerResourceMetric` is the correct one)

> That particular pod was in CrashLoopBackoff at the time and it wasn't until i removed it was it able to make scaling decisions
We're also running istio so there is another container in these pods and it's running ok; this may be why it's showing up in the metrics but unable to get info on the main container

Also, if you define the container level metric for each container, then the one for the healthy container should work. Although HPA wouldn't make any scale down, that's the correct and safe behaviour. (because we cannot know the scale down is safe or not since we have a crashing container.)

>  in my situation it was scaled to max and unable to scale down even though all metrics of the running pods showed it well under the targets
https://github.com/kubernetes/kubernetes/issues/127169#issuecomment-2333194320

Another part from the issue. In this case, HPA _shouldn't_ reduce the replica number because it may cause more disruption. 

Let's say there are 5 Pods with main-container and istio-proxy, two istio-proxy is OOM, while three istio-proxy containers are 10% resource utilization. How can we tell the scaling down is always safe in this situation? What if the scale down triggers other istio-proxy's OOM?
So, for me, it looks much safer to emit error and skip the scaling decision when missing container metric from any Pods.

**Comment by DP19** [comment]
Thanks for looking this over @sanposhiho! Definitely good things to think about here

> This part from the issue. Did you really use the container level metrics? AFAIK, the container level metric shouldn't emit this event. (FailedGetContainerResourceMetric is the correct one)

We were using container metrics, I think i originally misread some of the code here. From my view nothing was scaling even though we had hpa metrics for both containers within the pod and external metrics as well. So it still would have scaled up in this situation if any of the other metrics would have given that indication.

The Reason emitted  was `FailedGetResourceMetric` . We are running v1.29 in gke so they may have some downstream code that changed the reason but the error itself is coming from `getContainerMetrics`

excerpt:
```
  metrics:
  - containerResource:
      container: guides3
      name: cpu
      target:
        averageUtilization: 50
        type: Utilization
    type: ContainerResource
  - containerResource:
      container: istio-proxy
      name: cpu
      target:
        averageUtilization: 70
        type: Utilization
    type: ContainerResource
  - external:
      metric:
        name: istio.io|service|client|request_count
        selector:
          matchLabels:
            metric.labels.destination_canonical_revision: our-ervision
            metric.labels.destination_canonical_service_name: our-service
            metric.labels.destination_service_namespace: our-namespace
            resource.type: istio_canonical_service
      target:
        averageValue: "50"
        type: AverageValue
    type: External
```
> Also, if you define the container level metric for each container, then the one for the healthy container should work. Although HPA wouldn't make any scale down, that's the correct and safe behaviour. (because we cannot know the scale down is safe or not since we have a crashing container.)

Yeah that's what i expect as well looking through the code but the issue is this is the main container driving scale 99% of the time so not having these metrics considered at all could cause us not to scale up in a situation where these are overloaded and the other metrics look ok.

> Another part from the issue. In this case, HPA shouldn't reduce the replica number because it may cause more disruption.
>
> Let's say there are 5 Pods with main-container and istio-proxy, two istio-proxy is OOM, while three istio-proxy containers are 10% resource utilization. How can we tell the scaling down is always safe in this situation? What if the scale down triggers other istio-proxy's OOM?
So, for me, it looks much safer to emit error and skip the scaling decision when missing container metric from any Pods.

I think i would tend to agree in this situation, but it seems very, very unlikely to happen in a real world case; and if it was downscaling may not provide additional disruption.
In our case, we had an issue with setting up a Go ballast that was too large and OOM'ing the container on boot, not due to the load. Downscaling here would have been completely ok.

I think the larger issue is the more likely scenario, if we take the same example but instead of istio on the other containers being 10% utilization lets say they're 80% because they're getting their share of traffic that caused the other 2 to OOM and are now going to get even more of the load because those are un-ready. If none of the other metrics here that the HPA is configured for is telling the HPA to scale; we wouldn't increase scale in this situation and it would more likely than not that the other 3 containers would OOM.


I'm not entirely sure what the "right" answer here is to be honest; but in it's current form switching from Pod metrics to Container metrics does introduce this (in my view) unexpected behavior. Reading through the docs I would expect the container / pod metrics scalers to behave the same way, just differ in the exact metrics they're making decisions on.

Again, happy to have further discussions with other folks form sig-autoscaling on what the consensus might be here, thanks again for taking the time to review this

**Comment by sanposhiho** [comment]
So, my concern is mainly downscaling only actually.
Would both of us be happy if we would only allow scale-up and disallow scale-down when lacking metrics? 
Only allowing scale-up is aligned with other scenarios of HPA failure mode too, so I believe it'd make sense.

**Comment by DP19** [comment]
> So, my concern is mainly downscaling only actually.
Would both of us be happy if we would only allow scale-up and disallow scale-down when lacking metrics?
Only allowing scale-up is aligned with other scenarios of HPA failure mode too, so I believe it'd make sense.

Definitely aggree in the scale up situation and it should always make a decision here; however downscaling is actually the situation that brought me to find this bug. In one of our production workloads which does some batch job processing by looking at both cpu / number of items in a pubsub backlog. This workload uses istio, but for making requests to other workloads and isn't driven by request load so we moved it from Pod based to Container based HPA to ignore the isito pod metrics altogether as they never really change.

For ~8 hours during the day we noticed that the deployment was staying at max scale and it was due to missing metrics just for a single container out of the 160 or so running pods. These few pods throughout the day were OOM'ing but all other pods were fine and downscaling here would have been preferred over taking no action. This lead us to overspend by 2x that day and changing it back to a Pod based HPA got rid of this issue entirely

Expected / normal scale throughout the day with some OOMs (using Pod based HPA)
Screenshot 2024-09-15 at 11 14 22 AM

vs using contaier based HPA and the same ooms
Screenshot 2024-09-15 at 11 12 40 AM

cpu request utilization (which looks the same across both days from the scale screenshots)
Screenshot 2024-09-15 at 11 45 48 AM


We might want to consider the same proposed situation but under both the happy path and this new path:

> Let's say there are 5 Pods with main-container and istio-proxy, two istio-proxy is OOM, while three istio-proxy containers are 10% resource utilization. How can we tell the scaling down is always safe in this situation?

So under the happy path which is currently possible today, lets consider the 2 pods haven't OOM'd quite yet but are at 99% utilization and might be about to.
The overall utilization would be ~45% so if the target is above that we'd end up scaling down anyway, even though this may not be "safe" thing to do. This number also goes down as the scale grows and the number of pods under heavy utilization is the same, like it is in my case. There's periods of a half hour or more in our metrics where overall mean cpu utilization is at 0 and we're not scaling down due to a single pod that was Terminating and not reporting metrics.

With this, i wouldn't consider it to be a Failure mode at all. The pods are in a valid state (Terminating) and would end up getting filtered out of the calculation right after the return from this metrics collection.

With the new path, the OOM'd pods would be handled via `missingPods` in `calcPlainMetricReplicas` and if the situation would be to scale down they're counted as having exactly the utilization as the target ammount:
```
			// on a scale-down, treat missing pods as using exactly the target amount
			for podName := range missingPods {
				metrics[podName] = metricsclient.PodMetric{Value: targetUsage}
			}
```

In that case the utilization would be similar, lets say the Average target is 80% we'd end up with 38% utilization and it would scale down similar to the case where they were still running and heavily utilized.


All of this to say, i'd still like to at least advocate as best i can that we adopt this new behavior and consider this a bug. The Pod based HPA has been operating similarly since it's inception so many versions ago and hasn't been considered to adopt the behavior we have in the container based situation. 
If we do decide to only operate here based on the scale up case; we should also make the same adjustment in the Pod based HPA as i believe they should operate the same way (through a separate issue / pr 😅 )

**Comment by sanposhiho** [comment]
Still cannot agree about the downscaling. 
There are many possible scenarios that metrics are missing (e.g., somehow the metrics are temporarily unavailable in some Nodes), and it's not obvious that it's always safe to allow HPA to downscale the replicas in all of the scenarios. 
Even if, let's say, the down-scale when lacking metrics could benefit people in 80% scenarios safely, we shouldn't make it happen for the rest of 20% scenario. That's the most important point for me, in making the decision here.

---

cc @gjtempleton

need to include more people in the discussion.

**Comment by raywainman** [comment]
> The Reason emitted was FailedGetResourceMetric . We are running v1.29 in gke so they may have some downstream code that changed the reason but the error itself is coming from getContainerMetrics

This is exactly right, looks like there is bug in GKE where `FailedGetResourceMetrics` is emitted instead.

**Comment by DP19** [comment]
Sorry for yet another response, I know i've been pretty wordy here already going over different cases. Thanks again to everyone for taking the time to put eyes on this. In an attempt to be as succinct as possible for whomever looks at this next:

I'm viewing this as a bug because the `Resource` and `ContainerResouce` metric types _should_ behave the same; the code paths are nearly identical:
```
                 switch spec.Type
                           /\
                          /  \
                         /    \
                        /      \
  ResourceMetricSourceType    ContainerResourceMetricSourceType
             |                           |
             |                           |
             |                           |
computeStatusForResourceMetric   computeStatusForContainerResourceMetric
              \                         /               
               \                       /
                \                     /
                 \                   /
               computeStatusForResourceMetricGeneric
```

In the end they only differ in this single spot, depending on if we're looking for `container` metrics or not. The code paths are nearly identical 

```
	if container != "" {
		res = getContainerMetrics(ctx, metrics.Items, resource, container)
	} else {
		res = getPodMetrics(ctx, metrics.Items, resource)
	}
```

The error returned is one we make if we're missing any metrics; whereas `getPodMetrics` only logs this. Any hypotheticals we could compose should be answered just by looking at what we already do in the return of `GetResourceReplicas` which handles missing metrics and has been for the `Resource` metric type since around it's inception

**Comment by raywainman** [comment]
Right in `getPodMetrics` we simply omit the pods that have containers with missing information:

https://github.com/kubernetes/kubernetes/blob/e2c17c09a494f679e5a8fb46f2d5ca05363df0a8/pkg/controller/podautoscaler/metrics/client.go#L127-L133

In `getContainerMetrics` we abort with an error if we find a single pod with missing container information:

https://github.com/kubernetes/kubernetes/blob/e2c17c09a494f679e5a8fb46f2d5ca05363df0a8/pkg/controller/podautoscaler/metrics/client.go#L106-L108

In `GetResourceReplicas`, we count pods with missing metrics as `missing`:

https://github.com/kubernetes/kubernetes/blob/e2c17c09a494f679e5a8fb46f2d5ca05363df0a8/pkg/controller/podautoscaler/replica_calculator.go#L392-L397

Which in the calculator we treat in a special way:

https://github.com/kubernetes/kubernetes/blob/e2c17c09a494f679e5a8fb46f2d5ca05363df0a8/pkg/controller/podautoscaler/replica_calculator.go#L106-L120

The way I see it, if we don't return error in the metrics client when metrics are missing, the pod will then be labelled as `missing` and subject to the logic above.

@sanposhiho is this right? Would that address your concerns with the downscale logic?

Thanks for all the great input here folks!

File: pkg/controller/podautoscaler/metrics/client.go
Comment (sanposhiho) [inline pkg/controller/podautoscaler/metrics/client.go:104]:
```suggestion
			klog.FromContext(ctx).V(2).Info("Missing container metric", "container", container, "pod", klog.Kobj(pod))
```

Comment (sanposhiho) [reply pkg/controller/podautoscaler/metrics/client.go:104]: FYI - https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md

**Review by sanposhiho — COMMENTED** [review]
/unhold

Sorry for a long block, took time coming back here due to summer vacation.
I now got what I was missing, and I'm ok for `/unhold`ing. Still I think we can improve how we treat `missingPods` though, as @DP19 said, for now, we can just move the container level metrics to how we do with the pod level metrics.

@DP19 This PR requires a release note. Add something to describe a change briefly from the user perspective, please.

/assign @gjtempleton
for the approval.

File: pkg/controller/podautoscaler/metrics/client.go
Comment (DP19) [reply pkg/controller/podautoscaler/metrics/client.go]: We actually don't have the full `pod` resource here; just the `metricsapi.PodMetrics`. Updated it to use the same `"pod", klog.KRef(m.Namespace, m.Name)` that `getPodMetrics` is logging

**Review by sanposhiho — APPROVED** [review]
/lgtm

**Comment by gjtempleton** [comment]
Thanks for the discussion on this, agreed on being OK with this as standardising the behaviour between the two for now, though I agree with @sanposhiho that it highlights that the downscale behaviour in the case of missing metrics currently isn't as safe as it could be.
/approve

**Comment by gjtempleton** [comment]
@mwielgus can we get an approval from you as well? (Working on augmenting you as the sole reviewer for this.)

**Comment by DP19** [comment]
@mwielgus @gjtempleton anything else I can do to help get this over the line? Are we just waiting on the code freeze at this point?

**Review by soltysh — APPROVED** [review]
> Thanks for the discussion on this, agreed on being OK with this as standardising the behaviour between the two for now, though I agree with @sanposhiho that it highlights that the downscale behaviour in the case of missing metrics currently isn't as safe as it could be. /approve

/approve

**Pull request merged**
